### ISSUE TITLE
Add `cargo new --workspace`

### ISSUE THREAD

**Issue by camelid** [rust-lang/cargo#8365]
**Describe the problem you are trying to solve**

One has to manually create a new directory and setup a `Cargo.toml`, rather than being able to simply run `cargo new foo --workspace` and have a basic template set-up.

**Describe the solution you'd like**

Add the option `--workspace` to `cargo new` that creates a new workspace with the provided name.

**Notes**

The feature is not too important, but it seems like it would streamline the process of creating a new workspace a bit.

**Comment by pksunkara** [comment]
Just wanted to chime in and say that we have been looking to implement this with https://github.com/pksunkara/cargo-workspaces/pull/15. Once that PR is merged, you should be able to do `cargo ws init`.

**Comment by buzmeg** [comment]
Should this also be closed as postponed per #8411 ?

**Comment by Stargateur** [comment]
@buzmeg why is there another issue that ask about this feature ?

**Comment by H2rmone** [comment]
Same confusion here. In Javascript community, tools like lerna, yarn workspace and nx both have command to create workspace project. Cargo workspace includes many feature, I think it will save a lot of beginners' time if we have command to create workspace project like `cargo new --workspace`.

**Comment by dobromyslov** [comment]
Are there any plans to implement this feature? Looks like the `templates` feature was postponed last year.

**Comment by drebbe-intrepid** [comment]
Any update with this? I would love this feature also.

**Comment by epage** [comment]
I've clarified the title and linked out to #6378 since I always get these mixed up.

As for plans to move forward with this, someone would need to step up and work on a design for this which at least acknowledges the needs of #6378 so one doesn't block the other.

My gut feeling is there are enough use cases with various needs that we'd be better served by research into #5151.

**Comment by rosti-il** [comment]
I think it's worth to support the `--workspace` option not only in `cargo new` and `cargo init` but also in `cargo remove`.

**Comment by epage** [comment]
`cargo remove` automatically removes workspace dependencies when they are no longer used.

**Comment by rosti-il** [comment]
@epage you're right. I just got confused between `cargo new` and `cargo add` when thought about the opposite of `cargo remove`.

**Comment by snubwoody** [comment]
Any news on this? I use workspaces all the time and I think this would be a great feature. I would be willing to help if that's possible?

**Comment by vdudouyt** [comment]
I have the same question

**Comment by solidiquis** [comment]
+1

**Comment by jhpratt** [comment]
Any updates will be posted here. Please refrain from asking if there are any updates or similar.

**Comment by IzioDev** [comment]
It's ok, only 5 years passed. What is the blocker exactly?
Accepting contributions?

**Comment by epage** [comment]
@IzioDev if you read further up in the thread, there is an answer, see https://github.com/rust-lang/cargo/issues/8365#issuecomment-1567677063

**Comment by joshtriplett** [comment]
#6378 seems addressed at this point. I *think* that there's value in doing this independent of some template mechanism, for the same reason we have options to generate both binary and library crates. It doesn't need to handle every use case, just generate a useful enough workspace so that further additions of crates automatically get added to that workspace. The main thing I think we would need to decide is whether we generate a top-level workspace with or without a crate in the same manifest. And since we've generally been recommending against the former, it seems like we should generate the latter to make a better example for people.

**Comment by epage** [comment]
Should `cargo new --workspace` allow in-place updating the current package to also have a workspace?  Currently, we disallow that for `--lib` and `--bin`.

**Comment by epage** [comment]
A question I'm asking myself is "what value is `cargo new --workspace` getting us to motivate the design and support work for this new mode?

For a virtual package, we might literally be just dropping a `Cargo.toml` with
```toml
[workspace]
```

We could auto-set `resolver` so instead we get:
```toml
[workspace]
resolver = "3"
```

We could auto-set the inheritable edition if we assume everyone should have it (see also #15801), so instead we get:
```toml
[workspace]
resolver = "3"

[workspace.package]
edition = "2024"
```
(`cargo new` will automatically inherit it)

We could walk the current directory hierarchy and discover workspace members like `cargo workspace init` (pksunkara/cargo-workspaces#15) so we'd get
```toml
[workspace]
resolver = "3"
members = ["crates/foo", "crates/bar"]

[workspace.package]
edition = "2024"
```
**One downside** to this approach is whether we should encourage a style that allows for glob members.  It might be hard for `cargo workspace init` to infer when a glob could be used.  We could always include the glob `members = ["crates/*"]` but that could make `cargo` error.  Should we allow that?

If we walk the filesystem, should we still automatically setup `workspace.package.edition`?  Existing packages implies an existing set of policies while setting `workspace.package.edition = "<latest>"` implies more of a fresh start.

---

Is there anything else I'm missing that we could get out of this?

**Comment by trygveaa** [comment]
> Is there anything else I'm missing that we could get out of this?

Initializing a new VCS repository and creating a VCS ignore file, like `cargo new` does for packages.

**Comment by unsecretised** [comment]
*I* just wrote a zsh function if anyone wants it:
```zsh
# Add this to your ~/.zshrc
cargo-ws() {
  if [ $# -lt 1 ]; then
    echo "Usage: cargo-ws <workspace-name> [crate-names...]"
    return 1
  fi

  local workspace_name=$1
  shift
  local crates=("$@")

  mkdir -p "$workspace_name"
  cd "$workspace_name" || return 1

  # Create root Cargo.toml
  cat > Cargo.toml <<EOF
[workspace]
members = [
$(for c in "${crates[@]}"; do echo "    \"$c\","; done)
]
EOF

  # Add .gitignore
  cat > .gitignore <<EOF
/target
**/*.rs.bk
Cargo.lock
EOF

  # Create crates
  for c in "${crates[@]}"; do
    echo "Creating crate: $c"
    cargo new "$c" --bin
  done

  echo "✅ Workspace '$workspace_name' initialized!"
}
```

credits: chatgpt.com

**Comment by goddnsgit** [comment]
cargo add xxx1 xxx2 xxx3 时会给当前 toml里添加? 但toml是工作区的toml时,会出错,这其实很容易解决的吧?
我若为每个子crate 添加依赖的话,我为啥用这不成熟的工作区? 我分别创建文件夹不是更省事吗?

**Comment by vikaspoddar** [comment]
Is there any work being done for the command? Any PR?

**Comment by t1bur0n-rs** [comment]
Since `cargo new` inside an existing workspace already auto-updates the members list, the remaining gaps are:

1. Creating a **virtual workspace** (no root package) from scratch — there's no command for this today, only manual `Cargo.toml` creation
2. Adding an **existing crate** to a workspace — no way to do this without manually editing the root `Cargo.toml`
3. Creating a new member crate at a **custom path** from outside the workspace root

Two design directions to address all three gaps — sharing both to get team input before implementing:

---

**Option A — Extend `cargo new` and `cargo init`**

Add a `--workspace` flag to `cargo new` / `cargo init` for creating virtual workspaces, and a separate `--add-to-workspace` flag for adding existing crates which then automatically updates workspace toml:

```bash
# Create a virtual workspace from scratch
cargo new --workspace my-project
cargo new --workspace my-project --members api,core,cli

# Add an existing crate to a specified workspace
cargo new --add-to-workspace ./path/to/workspace ./some/existing/crate

# Create a new member at a custom path and add it
cargo new --lib core --path crates/core  # auto-adds to workspace if inside one
```

Pros: smaller surface area, familiar command, no new top-level subcommand to maintain.

---

**Option B — New `cargo workspace` subcommand**

A dedicated workspace lifecycle command with its own subcommands:

```bash
# Create a virtual workspace from scratch
cargo workspace init my-project
cargo workspace init my-project --members api,core,cli

# Add a new crate and register it in the workspace
cargo workspace add --bin api
cargo workspace add --lib core --path crates/core

# Add an existing crate to the workspace
cargo workspace add --existing ./some/existing/crate
```

Pros: cleaner separation of concerns, workspace lifecycle gets its own home, naturally extensible (`cargo workspace list`, `cargo workspace remove`, etc.), mirrors how other ecosystems handle this (yarn workspace, npm workspaces).

---

Both options address the same gaps. The `--existing` flag / subcommand is particularly useful — pointing an already-written crate into a workspace without rewriting anything, just updating the root `Cargo.toml` members list.

Happy to implement whichever direction the team prefers — just looking for design input before writing code.

**Comment by steffahn** [comment]
@t1bur0n-rs hi and thank you for your interest in contributing to Rust (specifically cargo). I’m a member of the moderation team, and we heard some concerns about your lastest (and first) contributions (i.e. your replies on the issue tracker): Several of your replies (& the commits) you’ve shared have some signs of being created - or at least edited - with the help of LLMs. We unfortunately don’t have a comperhensive policy to point to yet, but at least for the issue tracker, concensus generally is that we’d appreciate if you write all your messages here yourself.[^1]

[^1]: I hope you can understand our scepticism and please correct us if any suspicions raised were unfounded; it’s hard to tell these days *without looking really deeply into someone’s interactions* how much someone is just conveying information from some session with an LLM, or sharing their own thoughts just mangled through some LLM editing process.

If you’re interested in chatting with people in advance about design considerations, and in the process helping people to get a *slightly* better feel for who you are / how you’re approaching things, you could also open a thread [on our Zulip](https://rust-lang.zulipchat.com/) channel.

Looking at the commits from the branch you shared in [the other issue thread](https://redirect.github.com/rust-lang/cargo/issues/7322#issuecomment-4971661928), it looks to me you’re putting a bit much into each individual commit; e.g. combining a refactor *and* addition of new feature(s) all in one commit; though I’d leave the last words on judging this to the respective maintainers (e.g. of cargo, in this case).

**Comment by t1bur0n-rs** [comment]
@steffahn Hello. I will say the messages are as stated at the bottom me sharing my thoughts and attempting to describe my ideas and implementations and having the messages edited by an LLM but I want to emphasize that the code is my own I am not very good when it comes to formal writing and I think a lot of times I end up jumbling things together in a hard to read mess so I use the LLM to edit my messages so that it gets straight to the point cleanly. Again the code itself however is mine, the ideas are mine (at least I thought the idea of bulk install for cargo install was mine I did not know of the one issue that I was directed to that talked about it and I thought I had looked pretty well for issues about it) if there's concerns (which I expected there to be) about my lack of commits and repos on GitHub I would like to point out that this was my old GitHub during university and I have since switched to GitLab I only came back to this GitHub for sole purpose of wanting to contribute to cargo my GitLab is https://gitlab.com/t1bur0n. The same for this issue by the way the idea that I presented was entirely my own and I simply had an LLM proofread and edit the message to better clearly state what my intentions were. I will refrain from doing so in the future I was actually about to reply to ed page about design decisions for #7322 so I will not do that for my response. I will also open a thread on Zulip introducing myself so hopefully the ease of worry can be mitigated. I am sorry for any concerns that arose from this I will refrain from using LLMs for editing in the future and I again want to emphasize that the code is mine and I have no problem explaining my approach and/or thought process behind my implementations if there are any. the issue brough up with the commits in my opinion reflects why I do the editing to begin with because even with that it originally started as just a simple implementation for bulk installing crates but after reading other issues referenced I had went ahead and implemented those before fully getting feedback on the original issue (despite having said I wouldn't)
